edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/ServiceIdentityTree.cs

```diff
@@ -0,0 +1,275 @@
+// Copyright (c) Microsoft. All rights reserved.
+namespace Microsoft.Azure.Devices.Edge.Hub.Core
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Collections.Immutable;
+    using System.Linq;
+    using Microsoft.Azure.Devices.Edge.Hub.Core.Identity.Service;
+    using Microsoft.Azure.Devices.Edge.Util;
+    using Microsoft.Azure.Devices.Edge.Util.Concurrency;
+
+    /// <summary>
+    /// This is a tree implementation that uses the ParentScopes property
+    /// to reconstruct the hierarchy of modules/devices under the current
+    /// Edge device in a nested Edge environment.
+    /// </summary>
+    class ServiceIdentityTree : IAuthenticationChainProvider
+    {
+        public event EventHandler<string> ServiceIdentityRemoved;
+
+        readonly string rootDeviceId;
+        AtomicReference<ImmutableDictionary<string, ServiceIdentityTreeNode>> nodes;
+
+        public ServiceIdentityTree(string rootDeviceId)
```

Comment:
is this possible for a subtree's root device? if so, we should use a different name to avoid confusion of root device of the entire tree.